File: lab/python/textos/substitute.py
import re


def my_replace1(string, palavra_a_trocar, troca):
    return string.replace(palavra_a_trocar, troca)


def my_replace2(string, palavra_a_trocar, troca):
    new = ""
    string = string.split()
    for texto in string:
        if texto == palavra_a_trocar:
            new += troca + " "
        elif palavra_a_trocar in texto:
            for p in texto:
                if p == palavra_a_trocar:
                    new += troca + " "
                    break
                new += p
        else:
            new += texto + " "

    return new


def my_replace3(string, palavra_a_trocar, troca, ignore_case=False):
    flags = re.IGNORECASE if ignore_case else 0
    return re.sub(re.escape(palavra_a_trocar), troca, string, flags=flags)


print(my_replace1('Eu gosto de Python', "Python", "C"))
print(my_replace2('Eu gosto de Python', "Python", "C"))
print(my_replace3('Eu gosto de Python', "Python", "C"))
print()
print(my_replace1('Eu', "Eu", "Salve Cachorro"))
print(my_replace2('Eu', "Eu", "Salve Cachorro"))
print(my_replace3('Eu', "Eu", "Salve Cachorro"))
print()
print(my_replace1('Droga, errei aqui,', ",", " "))
print(my_replace2('Droga, errei aqui,', ",", " "))
print(my_replace3('Droga, errei aqui,', ",", " "))

File: lab/javascript/textos/substitute.js
function myReplace1(string, palavra, troca) {
  // troca todas as ocorrências (como str.replace do Python sem regex)
  return string.split(palavra).join(troca);
}

function myReplace2(string, palavra, troca) {
  // implementação manual sem regex (varrendo e montando o resultado)
  if (palavra.length === 0) return string; // evita loop infinito
  let out = "";
  for (let i = 0; i < string.length; ) {
    // se casa a partir de i
    if (string.startsWith(palavra, i)) {
      out += troca;
      i += palavra.length;
    } else {
      out += string[i];
      i++;
    }
  }
  return out;
}

function myReplace3(string, palavra, troca, ignore_case = false) {
  // regex seguro (escapa a palavra)
  const escape = (s) => s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  const flags = ignore_case ? "gi" : "g";
  const re = new RegExp(escape(palavra), flags);
  return string.replace(re, troca);
}

// ---- testes (equivalentes aos do Python)
console.log(myReplace1("Eu gosto de Python", "Python", "C"));
console.log(myReplace2("Eu gosto de Python", "Python", "C"));
console.log(myReplace3("Eu gosto de Python", "Python", "C"));
console.log();

console.log(myReplace1("Eu", "Eu", "Salve Cachorro"));
console.log(myReplace2("Eu", "Eu", "Salve Cachorro"));
console.log(myReplace3("Eu", "Eu", "Salve Cachorro"));
console.log();

console.log(myReplace1("Droga, errei aqui,", ",", " "));
console.log(myReplace2("Droga, errei aqui,", ",", " "));
console.log(myReplace3("Droga, errei aqui,", ",", " "));
